### DISCUSSION TITLE
Manually open/close Combobox (auto complete)

### DISCUSSION THREAD

**Discussion by xsjcTony** [tailwindlabs/headlessui#2300] [Ideas]
Just as simple as described in the title, maybe via a `ref` or something.

It's a little bit frustrating currently to achieve it

- If I use CSS to control it (`hidden` or `invisible`), it will result in clicking on options does nothing (value does not change)
- If I simulate an `click()` on the `Combox.Button` in `onFocus`, the dropdown will still remain open after I clicked an option by mouse

In conclusion one way always leads to some other issues, so I'm afraid I have to raise a feature request here to get the official support

**Comment by sarahdayan** [comment]
I needed to manually close the combobox because I had an item in there that wasn't part of the options, but a link at the bottom.

Initially I had the link placed inside `<ComboboxOptions>`, which caused the combobox to stay open when clicked.

I moved it outside (while controlling its display with an external wrapper + the `open` render prop) but clicking caused the combobox to immediately close, even before processing the `click` event.

I didn't want to rely on a `setTimeout` hack which would be quite unreliable. I ended up patching the library with [`patch-package`](https://www.npmjs.com/package/patch-package) to expose an `onClose` handler in render props.

- Relevant code section https://github.com/tailwindlabs/headlessui/blob/0b8deaf7a94124b0e78a92f0d63167f1180337ce/packages/%40headlessui-react/src/components/combobox/combobox.tsx#L898-L905
- Close handler https://github.com/tailwindlabs/headlessui/blob/0b8deaf7a94124b0e78a92f0d63167f1180337ce/packages/%40headlessui-react/src/components/combobox/combobox.tsx#L136

I'm not a fan of carrying patches around in projects because it keeps me from confidently updating dependencies, but if you really need this, it can do the job until an official solution comes up.

**Comment by augustl** [comment]
Agreed! Would be nice with some access to the internal machinery to make it possible to imperatively open the combobox. My specific use case is to create a custom combobox input component (used for a very custom multi select setup) and be able to open the combobox when my custom input receives focus, just like the builtin default input component does by invoking the internal machinery.
